### PR TITLE
Implement passive witnessing functionality

### PR DESCRIPTION
**Pull request by Tyratox**
For the following messages, the frontend will automatically send witness messages for the following object / action combinations:

- election#cast_vote
- chirp#add
- chirp#delete
- reaction#add

These are exactly the messages that are signed using pop tokens and whose content is determined by other people and hence should not require a manual verification.

https://github.com/dedis/popstellar/blob/dabd3f94ff5329781d5e564b4d73f3a608b76adb/fe1-web/src/core/network/jsonrpc/messages/MessageRegistry.ts#L55-L90

https://github.com/dedis/popstellar/blob/dabd3f94ff5329781d5e564b4d73f3a608b76adb/fe1-web/src/features/witness/network/messages/WitnessRegistry.ts#L35-L69

There is a feature toggle in place that prevents the actual network messages from being sent.

### CODE REVIEW

**Comment by lhmerino** [comment]
Perfect, sounds good.

**Review by lhmerino — DISMISSED** [review]
Looks great!

File: fe1-web/src/core/network/jsonrpc/messages/__tests__/MessageRegistry.test.ts
Comment (pierluca) [inline fe1-web/src/core/network/jsonrpc/messages/__tests__/MessageRegistry.test.ts:1]: why do you need to disable this ? are you trying to bypass the `private` field protection ? 

If so, a typical pattern is to create a `TestableMessageRegistry` which inherits from `MessageRegistry` and would somehow expose the ~~private~~ protected field. It makes the intention a tiny bit clearer.

File: fe1-web/src/core/network/jsonrpc/messages/MessageRegistry.ts
Comment (pierluca) [inline fe1-web/src/core/network/jsonrpc/messages/MessageRegistry.ts:125]: this feels very ad-hoc.

It's true that we could image a pre-processing and post-processing of `ProcessableMessage`, but I could imagine that we may have scenarios where a given pre-processing or post-processing applies to one message but not the other.

Taking a step back, do we really need something like this ? I'd opt for YAGNI 😅

File: fe1-web/src/features/witness/network/messages/WitnessRegistry.ts
Comment (pierluca) [inline fe1-web/src/features/witness/network/messages/WitnessRegistry.ts:1]: why duplicate this and not add a field in `MessageRegistry` ?

File: fe1-web/src/features/witness/network/index.ts
Comment (pierluca) [inline fe1-web/src/features/witness/network/index.ts:62]: here we're hooking up to message registry / message processing directly.
It feels a bit ad-hoc, even though the solution was developed to be generic.

what if instead we hooked up to the store and watched messages being marked as processed (e.g. the `processMessages` event in the `MessageReducer`) ? (as we discussed today for the election key problem)

I see more code / more logic as a clear disadvantage of what I'm saying, but it'd also be purely centralized in the `witness` feature and a feature couldn't, on its own, break the message registry logic.

as a completely different approach, we could also envision the `processMessages` action to be wrapped in a thunk, so that pre-processors and post-processors can be added at that level (in a completely transparent way to the message registry).

Any thoughts ? The current implementation is fine anyhow, I'm just wondering what would look better😄

File: fe1-web/src/core/network/jsonrpc/messages/MessageRegistry.ts
Comment (Tyratox) [reply fe1-web/src/core/network/jsonrpc/messages/MessageRegistry.ts:125]: I will re-implement it using redux 👍

File: fe1-web/src/core/network/jsonrpc/messages/__tests__/MessageRegistry.test.ts
Comment (Tyratox) [reply fe1-web/src/core/network/jsonrpc/messages/__tests__/MessageRegistry.test.ts:1]: Given that `afterProcessingHandlers` are removed anyway, this problem is solved as well :)
But yes, I agree. But it is not straight forward to do so, since I actually want to test the `MessageRegistry` and not some `TestableMessageRegistry`. Extending the class should work though 🤔

File: fe1-web/src/features/witness/network/messages/WitnessRegistry.ts
Comment (Tyratox) [reply fe1-web/src/features/witness/network/messages/WitnessRegistry.ts:1]: My thought was that it should be inside the feature and not part of the core?

File: fe1-web/src/features/witness/network/index.ts
Comment (Tyratox) [reply fe1-web/src/features/witness/network/index.ts:62]: Yes I think that is much nicer! I'll try to implement this using a store watcher now :)

File: fe1-web/src/core/keypair/__mocks__/KeyPairStore.ts
Comment (Xelowak) [inline fe1-web/src/core/keypair/__mocks__/KeyPairStore.ts:12]: Can't you write it like this? I feel like it's clearer and less verbose. Not sure about it though. 
```suggestion
  export const store = jest.fn<void, [KeyPair]>((kp: KeyPair) => {
    keyPair = kp;
  });
```
The same would apply to the function below and to every function where you called `.mockImplementation()`.

File: fe1-web/src/features/lao/functions/lao.ts
Comment (Xelowak) [inline fe1-web/src/features/lao/functions/lao.ts:28]: This feels a little bit repetitive.

File: fe1-web/src/features/lao/interface/Configuration.ts
Comment (Xelowak) [inline fe1-web/src/features/lao/interface/Configuration.ts:69]: Same comment.

File: fe1-web/src/features/witness/network/WitnessStoreWatcher.ts
Comment (Xelowak) [inline fe1-web/src/features/witness/network/WitnessStoreWatcher.ts:25]: Just a typo.
```suggestion
      // we have a witnessing entry for this message type
```

File: fe1-web/src/features/witness/network/messages/WitnessRegistry.ts
Comment (Xelowak) [reply fe1-web/src/features/witness/network/messages/WitnessRegistry.ts:1]: I agree that it's part of the feature, but on the other side it could easily fit in the `MessageRegistry`, as all features depend on this too. Maybe you can talk about it with @0xdieman.

File: fe1-web/src/features/witness/network/messages/WitnessRegistry.ts
Comment (Xelowak) [inline fe1-web/src/features/witness/network/messages/WitnessRegistry.ts:27]: Making a string-valued enum might help when debugging.

File: fe1-web/src/features/witness/network/messages/WitnessRegistry.ts
Comment (Xelowak) [inline fe1-web/src/features/witness/network/messages/WitnessRegistry.ts:71]: Is it really relevant to get the whole `WitnessEntry`? I think it would be better to get each part individually (so the `WitnessingType` only for now), so that you don't have to use the object once retrieved.

File: fe1-web/src/features/witness/network/messages/__tests__/WitnessRegistry.test.ts
Comment (Xelowak) [inline fe1-web/src/features/witness/network/messages/__tests__/WitnessRegistry.test.ts:14]: You're not actually testing anything there, right? For it to become relevant, you should add a check to the entries, but then it means that updating the registry would imply updating the test, which is not practical.

**Review by Xelowak — COMMENTED** [review]
Nice addition !
Thanks for all the documentation you added to your code, but I think that some of it is really too repetitive. When a function only returns a boolean, I think one sentence is enough to describe it, otherwise it becomes a little bit cumbersome.

File: fe1-web/src/features/witness/network/messages/WitnessRegistry.ts
Comment (Tyratox) [reply fe1-web/src/features/witness/network/messages/WitnessRegistry.ts:1]: 👍 I will do that! :)

File: fe1-web/src/core/keypair/__mocks__/KeyPairStore.ts
Comment (Tyratox) [reply fe1-web/src/core/keypair/__mocks__/KeyPairStore.ts:12]: Oh, indeed this is much nicer! I did not know about this, thank you! 😁

File: fe1-web/src/features/lao/functions/lao.ts
Comment (Tyratox) [reply fe1-web/src/features/lao/functions/lao.ts:28]: 👍

File: fe1-web/src/features/lao/interface/Configuration.ts
Comment (Tyratox) [reply fe1-web/src/features/lao/interface/Configuration.ts:69]: 👍

File: fe1-web/src/features/witness/network/messages/__tests__/WitnessRegistry.test.ts
Comment (Tyratox) [reply fe1-web/src/features/witness/network/messages/__tests__/WitnessRegistry.test.ts:14]: Well yes & no, the purpose of this test was to check whether it throws an error or not but I'll make this more explicit!

File: fe1-web/src/features/witness/network/index.ts
Comment (0xdieman) [inline fe1-web/src/features/witness/network/index.ts:27]: Did you leave this on purpose ?
Also, I'm not sure I understand the difference between those two, is it that if the handler fails, the entire message processing fails, when the subscribing way only "watches" (does not influence the processing logic even if it fails)?

**Review by 0xdieman — APPROVED** [review]

File: fe1-web/src/features/witness/network/index.ts
Comment (Tyratox) [reply fe1-web/src/features/witness/network/index.ts]: Oh whoops no, thank you!

Both approaches would have not make the message processing fail, the watcher approach is just a bit more "clean" since it directly reacts to state changes (when messages are added) and not to network events :)

For the other approach I had to add an additional `addAfterProcessingHandler` which is no longer needed with the store watcher :)

**Review by Xelowak — COMMENTED** [review]
Thanks a lot for all the changes! Since I don't know if you made a decision about the WitnessRegistry, I'll let you merge this branch when ready.

**Review by Xelowak — APPROVED** [review]

**Pull request merged**
